dwave/optimization/include/dwave-optimization/nodes/constants.hpp

```diff
@@ -29,6 +29,25 @@ namespace dwave::optimization {
 /// A contiguous block of numbers.
 class ConstantNode : public ArrayOutputMixin<ArrayNode> {
  public:
+    struct DataSource {
+        DataSource() = default;
+        virtual ~DataSource() = default;
+        // Technically should satisfy the rule of five here, and it's good
+        // to be on the safe side by deleting the unnecessary constructors
+        DataSource(const DataSource&) = delete;
+        DataSource& operator=(const DataSource&) = delete;
+        DataSource(DataSource&& other) noexcept = default;
+        DataSource& operator=(DataSource&& other) noexcept = delete;
+    };
+
+    struct OwningDataSource : public DataSource {
```

Comment:
I think the one can be private

Reply:
Fixed in https://github.com/dwavesystems/dwave-optimization/pull/337/commits/e46e450b5e532b31dc93ff36846915a120c22f05